File: 03- Subprogramas y recursividad/recursividad/java/Recursividad1.java
/**
 * Descripción: Imprimir del 1 al n utilizando la recursividad.
 * Autor: Santiago Nicolás De la mora Núñez
 * Fecha de creación: 10/03/2026
 */

public class Recursividad1 {
    public static void main(String[] args) {
        imprimir(3);

        System.out.println();

        imprimirHaciaAtras(3);
    }

    /*
    imprimir(3)
        |
        imprimir(3 - 1)
        sout(3)
            |
            imprimir(2 - 1)
            sout(2)
                |
                imprimir(1 - 1) se detiene
                sout(1)
                    |
                    return  (se detiene)
     */
    private static void imprimir(int n) {
        // Caso base
        if (n == 0) {
            return;
        }

        // Caso general (recursivo)
        else {
            imprimir(n - 1);

            System.out.print(n + " ");
        }
    }

    /*
    imprimirHaciaAtras(3)
        |
        sout(3)
        imprimirHaciaAtras(3 - 1)
            |
            sout(2)
            imprimirHaciaAtras(2 - 1)
                |
                sout(1)
                imprimirHaciaAtras(1 - 1)
                    |
                    return  (se detiene)
     */
    private static void imprimirHaciaAtras(int n) {
        // Caso base
        if (n == 0) {
            return;
        }

        // Caso general (recursivo)
        else {
            System.out.print(n + " ");
            imprimirHaciaAtras(n - 1);
        }
    }
}

File: 03- Subprogramas y recursividad/recursividad/python/recursividad1.py
"""
Descripción: Imprimir del 1 al n utilizando la recursividad.
Autor: Santiago Nicolás De la mora Núñez
Fecha de creación: 11/04/2026
"""

def main():
    imprimir(3)

    print()

    imprimir_hacia_atras(3)

'''
imprimir(3)
    |
    imprimir(3 - 1)
    sout(3)
        |
        imprimir(2 - 1)
        sout(2)
            |
            imprimir(1 - 1) se detiene
            sout(1)
                |
                return  (se detiene)
'''
def imprimir(n):
    # Caso base
    if n == 0:
        return
    
    # Caso general (recursivo)
    else:
        imprimir(n - 1)
        print(f"{n}", end=' ')

'''
imprimir_hacia_atras(3)
    |
    sout(3)
    imprimir_hacia_atras(3 - 1)
        |
        sout(2)
        imprimir_hacia_atras(2 - 1)
            |
            sout(1)
            imprimir_hacia_atras(1 - 1)
                |
                return  (se detiene)
'''
def imprimir_hacia_atras(n):
    # Caso base
    if n == 0:
        return
    
    # Caso general (recursivo)
    else:
        print(f"{n}", end=' ')
        imprimir(n - 1)

# Ejecuta el programa
if __name__ == '__main__':
    main()